home/home4_1/Main.java

```diff
@@ -0,0 +1,127 @@
+        predefinedList.put("Special", new String[]{"tomatoes", "ketchup", "mozzarella", "mayo"});
+    }
+}
+
+public class Main {
+
+    public static Map<String, Double> readProducts(String filename) {
+        try (BufferedReader reader = new BufferedReader(new FileReader(filename))) {
+            String line;
+            while ((line = reader.readLine()) != null) {
+                String[] parts = line.split("-");
+                if (parts.length == 2) {
+                    String name = parts[0].trim();
+                    double price = Double.parseDouble(parts[1].trim());
+                    Constants.productList.put(name, price);
+                }
+            }
+        } catch (Exception e) {
+            System.err.println("An unexpected error occurred: " + e.getMessage());
+            e.printStackTrace();
+        }
+        return Constants.productList;
+    }
+
+    public static Optional<String> ingredientInputer(String message, Map<String, Double> availableProducts, Scanner scanner) {
+        System.out.println("Pizza builder " + message);
+        String product = scanner.nextLine().trim();
+        if (availableProducts.containsKey(product)) {
+            return Optional.of(product);
+        } else {
+            return Optional.empty();
+        }
+    }
+
+    public static Optional<String[]> buildPredef(Scanner scanner) {
+        System.out.println("Here are predefined pizzas:");
+        System.out.println("Carbonara (ham, mozzarella, onion, onion)");
+        System.out.println("Hawaii (ham, mozzarella, mozzarella, mayo)");
+        System.out.println("Special (tomatoes, ketchup, mozzarella, mayo)");
+        System.out.println("Make your choice:");
+
+        String pizza = scanner.nextLine().trim();
+        if (Constants.predefinedList.containsKey(pizza)) {
+            return Optional.of(Constants.predefinedList.get(pizza));
+        } else {
+            return Optional.empty();
+        }
+    }
+
+    public static Optional<String[]> buildCustom(Scanner scanner) {
+        String[] ingredients = new String[4];
+
+        for (Map.Entry<String, Double> product : Constants.productList.entrySet()) {
+            System.out.println(product.getKey() + "-" + product.getValue() + "$");
+        }
+
+        Constants.productList.put("nothing", 0.0);
+
+        ingredients[0] = ingredientInputer("What first do you want", Constants.productList, scanner).orElse("nothing");
+        ingredients[1] = ingredientInputer("What second do you want", Constants.productList, scanner).orElse("nothing");
+        ingredients[2] = ingredientInputer("What third do you want", Constants.productList, scanner).orElse("nothing");
+        ingredients[3] = ingredientInputer("What fourth do you want", Constants.productList, scanner).orElse("nothing");
+
+        return Optional.of(ingredients);
+    }
+
+    public static void main(String[] args) {
```

Comment:
This function seems to be long enough. Maybe we can split it to smaller parts - some functions or separate class to handle user interaction scenarios. In general we have top level menu predefined/custom. Therefore details of custom pizza should be on other level of abstraction for example wrapped into some handler function. It should be organised like
case predefined => handlePredefined();
case custom => handleCustom();
case other => handleErrors();